tests/unit/test_middleware/test_constraints.py

```diff
@@ -0,0 +1,365 @@
+    two_two = MiddlewareTwo()
+    two_one.constraints = two_two.constraints = MiddlewareConstraints(**{constraint: (MiddlewareOne,)})  # type: ignore[arg-type]
+
+    given = (
+        two_one,
+        one,
+        two_two,
+    )
+
+    with pytest.raises(ConstraintViolationError, match=f"MiddlewareTwo.*{constraint}.*MiddlewareOne"):
+        check_middleware_constraints(given)
+
+
+@pytest.mark.parametrize("constraint", ["before", "after"])
+def test_order_multiple_instances_referee(constraint: str) -> None:
+    # check priority works in all cases if more than one instances of the referenced
+    # middleware (i.e. the one referenced in the priority) is present
+    one_one = MiddlewareOne()
+    one_two = MiddlewareOne()
+    two = MiddlewareTwo()
+    two.constraints = MiddlewareConstraints(**{constraint: (MiddlewareOne,)})  # type: ignore[arg-type]
+
+    given = (
+        one_one,
+        two,
+        one_two,
+    )
+
+    with pytest.raises(ConstraintViolationError, match=f"MiddlewareTwo.*{constraint}.*MiddlewareOne"):
+        check_middleware_constraints(given)
+
+
+def test_order_subclass() -> None:
+    # if a base class of a middleware is referenced, its priority should apply to all
+    # subclasses
+    class SubclassOfOne(MiddlewareOne):
+        pass
+
+    class CustomBaseClass:
+        pass
+
+    class CustomSubClass(CustomBaseClass):
+        pass
+
+    subclass_of_one = SubclassOfOne()
+    one = MiddlewareOne()
+    two = MiddlewareTwo()
+    three = MiddlewareThree()
+    custom_sub_class = CustomSubClass()
+
+    two.constraints = MiddlewareConstraints(before=(MiddlewareOne,))
+    three.constraints = MiddlewareConstraints(after=(CustomBaseClass,))  # type: ignore[arg-type]
+
+    given = (
+        one,
+        subclass_of_one,
+        two,
+        three,
+        custom_sub_class,
+    )
+
+    # expect the first constraint to be violated
+    with pytest.raises(ConstraintViolationError, match="MiddlewareTwo.*before.*MiddlewareOne"):
+        check_middleware_constraints(given)  # type: ignore[arg-type]
+
+
+def test_order_handle_cycle() -> None:
+    one = MiddlewareOne()
+    two = MiddlewareTwo()
+    one.constraints = MiddlewareConstraints(before=(MiddlewareTwo,))
+    two.constraints = MiddlewareConstraints(before=(MiddlewareOne,))
+
+    # TODO: Handle cycle error to point to specific instances
+    with pytest.raises(CycleError):
+        check_middleware_constraints((one, two))
+
+
+def test_order_handle_self_referential() -> None:
+    one = MiddlewareOne()
+    one.constraints = MiddlewareConstraints(before=(MiddlewareOne,))
+
+    with pytest.raises(CycleError):
+        check_middleware_constraints((one,))
+
+
+def test_first() -> None:
+    one = MiddlewareOne()
+    two = MiddlewareTwo()
+    one.constraints = MiddlewareConstraints(first=True)
+
+    with pytest.raises(
+        ConstraintViolationError, match="MiddlewareOne.*must be at the top of the stack. Found at index 1"
```

Comment:
I'd rather see `first` instead of `top of the stack` but I guess it's just a matter of taste